### PR TITLE
[DEV] Feature branch to re-route flow to illumina device tables.

### PR DESCRIPTION
**Pull request by ChrOertlin**
# Description

Feature branch to re-route the post-processing database flow from the `Flowcell` -> `SampleLaneSequencingMetrics` model paradigm to `IlluminaFlowCell` -> `IlluminaSequencingRun` -> `IlluminaSampleSequencingMetrics`. 
The choice to use a feature branch is to collect all changes and make sure they are tested. Since we need to alter core logic that requires a database migration we cannot do these step by step.

PR in servers: https://github.com/Clinical-Genomics/servers/pull/1394
Data migration script: https://github.com/Clinical-Genomics/add-new-tech/issues/7

## Deploy
```shell
bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b develop-illumina-devices -a
```

## CLI command changes
Commands that need to be updated in servers

- [x] `cg clean flow-cells` -> `cg clean illumina-runs` (requires servers update, PR: <placeholder>)
- [x] `cg backup encrypt-flow-cells` -> `cg backup encrypt-illumina-runs` (requires servers update, PR: <placeholder>)
- [x] `cg backup flow-cells` -> `cg backup illumina-runs` (requires servers update, PR: <placeholder>)
- [x] `cg backup fetch-flow-cell`-> `cg backup fetch-illumina-run` (no update needed)
- [x] `cg demultiplex finish flow-cell` -> `cg demultiplex finish illumina-run` (no update needed in servers
- [x] `cg get flow-cell` -> `cg get sequencing-run` (Does not require a PR in servers)
- [x] `cg set flowcell` -> `cg set sequencing-run` (Does not require a PR in servers)
- [x] `cg store flow-cell`-> `cg store illumina-run` (Does not require a PR in servers)
- [x] `cg store demultiplexed-flow-cell` -> `cg store demultiplexed-run` (Does not require a PR in servers)
- [x] `cg decompress flow-cell` -> `cg decompress illumina-run` (Does not require a PR in servers)
https://github.com/Clinical-Genomics/servers/pull/1394
## Tests
CO:
- [x] NIPT 
  - [x] `cg -l DEBUG upload nipt case healthycollie --dry-run`
  - [x] Requirements: A NIPT case present on stage - Healthy collie
- [x] Clean illumina runs
  - [x] `cg clean illumina-runs --dry-run`
  - [x] Requires: Illumina runs present on `/home/proj/stage/sequencing_data/illumina/sequencing_runs` and in statusDB
- [x] Encrypt illumina runs 
  - [x] `cg backup encrypt-illumina-runs --dry-run`
  - [x] Requires: sequencing-runs in `/home/proj/stage/sequencing_data/illumina/sequencing_runs`
- [x] backup illumina runs
  - [x] `cg backup illumina-runs --dry-run`
  - [x]  Requires: sequencing-runs in `/home/proj/stage/sequencing_data/illumina/sequencing_runs`
- [x] `cg demultiplex finish illumina-run`
  - [x] Requires: sequencing-runs in `/home/proj/stage/sequencing_data/illumina/demultiplexed-runs` and sequencing-runs in `/home/proj/stage/sequencing_data/illumina/sequencing_runs`
  - [x] Rename the fastq files in the demux directory removing the flow cell id prefix
  - [x] Remove flow cells from statusdb and housekeeper prior to running the command
  - [x] Check that the validation protocol fulfilled.
-----------------------------------------
SD:

- [x] Fetching information from an Illumina run and a sample from statusdb
  - [x] Commands to test:
	- [x] `cg get sequencing-run <flow-cell-id>`
	- [x] `cg get sequencing-run <flow-cell-id> --samples`
	- [x] `cg get sample --hide-flow-cell/<no flag>` 
  - [x] Requirements:
	- [x] A post-processed run with samples (any) in statusDB
	- [x] A sample from a post-processed flow cell in statusDB
- [x] Modify the data availability status of a sequencing run from statusDB
  - [x] Commands to test
    - [x] `cg set sequencing-run <flow-cell-id> --data-availability <some_data_status>`
  - [x] Requirements 
    - [x] An Illumina run in statusDB
- [x] Retrieve spring files
  - [x] Commands to run
    - [x] `cg backup retrieve-spring-files --case-id <case_id> --dry-run`
    - [x] `cg backup retrieve-spring-files --sample-id <sample_id> --dry-run`
    - [x] `cg backup retrieve-spring-files --flow-cell-id <flow cell_id> --dry-run`
  - [x] Requirements
    - [x] A flow cell with samples in statusDB
    - [x] A case with samples in statusDB
    - [x] A sample in statusDB
- [x] Fetch an Illumina run that is going to be backed up
  - [x] Commands to test:
    - [x] `cg backup fetch-illumina-run --dry-run`
    - [x] `cg backup fetch-illumina-run <flow_cell_id> --dry-run`
  - [x] Requirements
    - [x] A sequencing run in statusDB
- [x] Ensure all flow cells ar eon disk for a mip-dna case
  - [x] Command to test
    - [x] `cg workflow mip-dna ensure_illumina_runs_on_disk <case_id>`
  - [x] Requirements
    - [x] A MIP-DNA case (`livingox`) with all its Illumina sequencing runs on disk (`HHT5NDSXX`)
- [x] Deploy in the VM or locally
  - [x] Commands to test
    - [x] Run cg-API stage locally
  - [x] Requirements
    - [x] Have Illumina flow cells, Illumina sequencing runs and sample sequencing metrics in statusDB. (Run inhouse script `migrate_illumina_tables.py`)  

# Test log

## NIPT upload
Success!
```
----------------- UPLOAD -----------------
*** NIPT UPLOAD START ***
Set dry run to True
Sequencing run for HTJ2MDRX3 passed QC for case healthycollie.
*** Statina UPLOAD START ***
Set dry run to True
Fetch latest version from bundle healthycollie
Fetching files with tags in [nipt,metrics]
Fetching files from version 169908
Fetch latest version from bundle healthycollie
Fetching files with tags in [nipt,multiqc-html]
Fetching files from version 169908
Fetch latest version from bundle healthycollie
Fetching files with tags in [nipt,wisecondor]
Fetching files from version 169908
Found file paths for statina upload: {"result_file":"/home/proj/stage/housekeeper-bundles/healthycollie/2023-12-19/2325531_NIPT.csv","multiqc_report":"/home/proj/stage/housekeeper-bundles/healthycollie/2023-12-19/multiqc_report.html","segmental_calls":"/home/proj/stage/housekeeper-bundles/healthycollie/2023-12-19"}
Set dry run to True
Sequencing run for HTJ2MDRX3 passed QC for case healthycollie.
*** NIPT FTP UPLOAD START ***
Fetch latest version from bundle healthycollie
Fetching files with tags in [nipt,metrics]
Fetching files from version 169908
Results file found: /home/proj/stage/housekeeper-bundles/healthycollie/2023-12-19/2325531_NIPT.csv
Starting ftp upload!
Would upload results file to ftp server: /home/proj/stage/housekeeper-bundles/healthycollie/2023-12-19/2325531_NIPT.csv
Upload ftp finished!
healthycollie: analysis uploaded!
```

## cg demultiplex samplesheet create

```
[christian.oertlin@hasta:~] [S_base] $ cg demultiplex samplesheet create 20231108_LH00188_0028_B22F52TLT3
Running cg demultiplex.
Getting a valid sample sheet for flow cell 20231108_LH00188_0028_B22F52TLT3
Called undefined __fields__ on HousekeeperAPI, please wrap
Sample sheet file for flowcell 22F52TLT3 not found in Housekeeper!
Sample sheet from Housekeeper is not correctly formatted or does not exist, trying sample sheet in sequencing directory
No index settings found in sample sheet
Sample sheet from sequencing directory is not correctly formatted or does not exist, creating new sample sheet
Fetching samples from lims for flowcell 22F52TLT3
Constructing sample sheet for the novaseqx flow cell 22F52TLT3
Updating barcode mismatch values for samples in lane 1
Updating barcode mismatch values for samples in lane 2
Updating barcode mismatch values for samples in lane 3
Updating barcode mismatch values for samples in lane 4
Updating barcode mismatch values for samples in lane 5
Updating barcode mismatch values for samples in lane 6
Updating barcode mismatch values for samples in lane 7
Updating barcode mismatch values for samples in lane 8
Creating sample sheet content
Samplesheet passed validation
Writing sample sheet to /home/proj/stage/sequencing_data/illumina/sequencing-runs/20231108_LH00188_0028_B22F52TLT3/SampleSheet.csv
Sample sheet file for flowcell 22F52TLT3 not found in Housekeeper!
Created new bundle version dir: /home/proj/stage/housekeeper-bundles/22F52TLT3/2023-11-08
Linked file: /home/proj/stage/sequencing_data/illumina/sequencing-runs/20231108_LH00188_0028_B22F52TLT3/SampleSheet.csv -> /home/proj/stage/housekeeper-bundles/22F52TLT3/2023-11-08/SampleSheet.csv
File added to Housekeeper bundle 22F52TLT3
```
### CLI command suite is updated
image
### `cg demultiplex finish illumina-run`
- [x] Data added to relevant table in statusdb
- [x] Files added to housekeeper

## cg demultiplex sequencing-run --dry-run
Works like a charm

## cg clean illumina-runs --dry-run
image

Correctly identifies that sequencing runs are not there when 
Also identifies sequencing-runs that should be cleaned

## cg backup encrypt-illumina-runs
Succes with a dry run
image
 
## cg backup illumina-runs --dry-run

Success with a dry run

image

### CODE REVIEW

**Comment by ChrOertlin** [comment]
# Tests

## cg demultiplex samplesheet create

```
[christian.oertlin@hasta:~] [S_base] $ cg demultiplex samplesheet create 20231108_LH00188_0028_B22F52TLT3
Running cg demultiplex.
Getting a valid sample sheet for flow cell 20231108_LH00188_0028_B22F52TLT3
Called undefined __fields__ on HousekeeperAPI, please wrap
Sample sheet file for flowcell 22F52TLT3 not found in Housekeeper!
Sample sheet from Housekeeper is not correctly formatted or does not exist, trying sample sheet in sequencing directory
No index settings found in sample sheet
Sample sheet from sequencing directory is not correctly formatted or does not exist, creating new sample sheet
Fetching samples from lims for flowcell 22F52TLT3
Constructing sample sheet for the novaseqx flow cell 22F52TLT3
Updating barcode mismatch values for samples in lane 1
Updating barcode mismatch values for samples in lane 2
Updating barcode mismatch values for samples in lane 3
Updating barcode mismatch values for samples in lane 4
Updating barcode mismatch values for samples in lane 5
Updating barcode mismatch values for samples in lane 6
Updating barcode mismatch values for samples in lane 7
Updating barcode mismatch values for samples in lane 8
Creating sample sheet content
Samplesheet passed validation
Writing sample sheet to /home/proj/stage/sequencing_data/illumina/sequencing-runs/20231108_LH00188_0028_B22F52TLT3/SampleSheet.csv
Sample sheet file for flowcell 22F52TLT3 not found in Housekeeper!
Created new bundle version dir: /home/proj/stage/housekeeper-bundles/22F52TLT3/2023-11-08
Linked file: /home/proj/stage/sequencing_data/illumina/sequencing-runs/20231108_LH00188_0028_B22F52TLT3/SampleSheet.csv -> /home/proj/stage/housekeeper-bundles/22F52TLT3/2023-11-08/SampleSheet.csv
File added to Housekeeper bundle 22F52TLT3
```
## cg demultiplex finish illumina-run 
### CLI command suite is updated
image
### `cg demultiplex finish illumina-run`
- [ ] Data added to relevant table in statusdb
- [ ] Files added to housekeeper

**Review by diitaz93 — APPROVED** [review]
All tests passing ⭐

**Comment by diitaz93** [comment]
To be merged on 2024-07-10

**Comment by diitaz93** [comment]
- [ ] Deployed to stage:
```shell
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
fatal: Not a git repository (or any parent up to mount point /home)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
repository is clean
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/js.diazboada
done.
Log deploy... done.
cg, version 62.0.0
[js.diazboada@hasta:~] [S_base] $ up
```
- [ ] Deployed to production:
```shell
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
[js.diazboada@hasta:~] [P_base] $ 
```

File: cg/meta/clean/clean_flow_cells.py [hunk 1]
```diff
@@ -1,184 +0,0 @@
...
-        Raises:
-            CleanFlowCellFailedError when any exception is caught
-        """
-        try:
-            self.set_sample_sheet_path_from_housekeeper()
-            if self.can_flow_cell_directory_be_deleted():
-                if self.dry_run:
-                    LOG.debug(f"Dry run: Would have removed: {self.flow_cell.path}")
-                    return
-                remove_directory_and_contents(self.flow_cell.path)
```
Comment (Vince-janv) [inline cg/meta/clean/clean_flow_cells.py:61]: Here 👋

File: cg/services/illumina_services/cleaning_services/clean_runs_service.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+            self.set_sample_sheet_path_from_housekeeper()
+            if self.can_run_directory_be_deleted():
+                if self.dry_run:
+                    LOG.debug(
+                        f"Dry run: Would have removed: {self.sequencing_run_dir_data.get_sequencing_runs_dir()}"
+                    )
+                    return
+                remove_directory_and_contents(
+                    self.sequencing_run_dir_data.get_sequencing_runs_dir()
+                )
```
Comment (Vince-janv) [inline cg/services/illumina_services/cleaning_services/clean_runs_service.py:74]: This is a change in logic compared to https://github.com/Clinical-Genomics/cg/pull/3349/files#r2011986932

**Pull request merged**